internal/cloud/backend_plan.go

```diff
@@ -95,7 +87,25 @@ func (b *Cloud) opPlan(stopCtx, cancelCtx context.Context, op *backend.Operation
 		return nil, diags.Err()
 	}
 
-	return b.plan(stopCtx, cancelCtx, op, w)
+	run, err := b.plan(stopCtx, cancelCtx, op, w)
+
+	// Save plan even if the run errored, as long as it actually exists
+	if run != nil && run.ID != "" && op.PlanOutPath != "" {
+		bookmark := cloudplan.NewSavedPlanBookmark(run.ID, b.hostname)
+		saveErr := bookmark.Save(op.PlanOutPath)
+		// Maybe combine errors
+		if err != nil && saveErr != nil {
+			err = fmt.Errorf("%w\nAdditionally, an error occurred when saving the plan: %s", err, saveErr.Error())
+		} else if err == nil {
+			err = saveErr
+		}
+	}
+
+	// Cloud integration currently doesn't support genconfig, but we might have
+	// saved a cloud plan file.
+	op.View.PlanNextStep(op.PlanOutPath, "")
+
+	return run, err
```

Comment:
Overall this logic confuses me a bit, so I'm hoping to gather more context here:

Why do we attempt to save the plan if the run fails which can potentially be in the planning phase? Seems like Terraform would have to know the status of the run, i.e a run that is in a state later than `planned` and only save the plan then. 

This is how I imagined the logic would look like in mostly pseudcode:
```go
r, err := b.plan()
if err != nil {
  saveErr := attemptToSavePlan()
  return r, errors.Join(err, saveErrr)
}
attemptToSavePlan()
// ... other stuff
}

// and then
func attemptToSavePlan(r *tfe.Run) error {
  // ensures the plan did succeed
  // this doesn't have to be a separate method
  if !validRunStatus() {
    // we don't need to pile on another error here, since that would be captured by the original error in `b.plan()`
    return nil
  }
  
  // do the plan savin' stuff
}
```

Reply:
Well, my reasoning was that the user might want to inspect the plan with `terraform show`, even if it failed. 

That said... the usefulness with a failed plan is minimal:

![image](https://github.com/hashicorp/terraform/assets/484309/b9d26279-6e3f-4d93-b62b-9c549f9de3d9)

You get a link to click to view the run in the UI, and an error message, and that's about it. And if you _just now_ performed an errored plan, you've already got the link in your terminal from that. 

So, I could be convinced to abort the save. 🤔

Reply:
Yeah okay changed it to not do that 👍🏼